plugins/inputs/aliyuncms/aliyuncms.go

```diff
@@ -0,0 +1,597 @@
+		requestDimensionsStr string              //String representation of the above
+
+	}
+
+	// Dimension describe how to get metrics
+	Dimension struct {
+		Value string `toml:"value"`
+	}
+
+	aliyuncmsClient interface {
+		DescribeMetricList(request *cms.DescribeMetricListRequest) (response *cms.DescribeMetricListResponse, err error)
+	}
+)
+
+// SampleConfig implements telegraf.Inputs interface
+func (s *AliyunCMS) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description implements telegraf.Inputs interface
+func (s *AliyunCMS) Description() string {
+	return description
+}
+
+func (s *AliyunCMS) Init() error {
+
+	if s.Project == "" {
+		return errors.New("project is not set")
+	}
+
+	var (
+		roleSessionExpiration = 600
+		sessionExpiration     = 600
+	)
+	configuration := &providers.Configuration{
+		AccessKeyID:           s.AccessKeyID,
+		AccessKeySecret:       s.AccessKeySecret,
+		AccessKeyStsToken:     s.AccessKeyStsToken,
+		RoleArn:               s.RoleArn,
+		RoleSessionName:       s.RoleSessionName,
+		RoleSessionExpiration: &roleSessionExpiration,
+		PrivateKey:            s.PrivateKey,
+		PublicKeyID:           s.PublicKeyID,
+		SessionExpiration:     &sessionExpiration,
+		RoleName:              s.RoleName,
+	}
+	credentialProviders := []providers.Provider{
+		providers.NewConfigurationCredentialProvider(configuration),
+		providers.NewEnvCredentialProvider(),
+		providers.NewInstanceMetadataProvider(),
+	}
+	credential, err := providers.NewChainProvider(credentialProviders).Retrieve()
+	if err != nil {
+		return errors.Errorf("failed to retrieve credential: %v", err)
+	}
+	s.client, err = cms.NewClientWithOptions("", sdk.NewConfig(), credential)
+	if err != nil {
+		return errors.Errorf("failed to create cms client: %v", err)
+	}
+
+	//check metrics dimensions consistency
+	for _, metric := range s.Metrics {
+		if metric.Dimensions != "" {
+			metric.dimensionsUdObj = map[string]string{}
+			metric.dimensionsUdArr = []map[string]string{}
+			err := json.Unmarshal([]byte(metric.Dimensions), &metric.dimensionsUdObj)
+			if err != nil {
+				err := json.Unmarshal([]byte(metric.Dimensions), &metric.dimensionsUdArr)
+				return errors.Errorf("Can't parse dimensions (it is neither obj, nor array) %q :%v", metric.Dimensions, err)
+			}
+		}
+	}
+
+	s.measurement = formatMeasurement(s.Project)
+
+	//Init discovery...
+	if s.dt == nil { //Support for tests
+		s.dt, err = NewDiscoveryTool(s.DiscoveryRegions, s.Project, credential, int(float32(s.RateLimit)*0.2), s.DiscoveryInterval.Duration)
+		if err != nil {
+			lg.logE("Discovery tool is not activated: %v", err)
+			s.dt = nil
+			return nil
+		}
+	}
+
+	s.discoveryData, err = s.dt.getDiscoveryDataAllRegions(nil)
+	if err != nil {
+		lg.logE("Discovery tool is not activated: %v", err)
+		s.dt = nil
+		return nil
+	}
+
+	lg.logI("%d object(s) discovered...", len(s.discoveryData))
+	//Start periodic discovery process
+	s.dt.Start()
```

Comment:
this shouldn't happen here, it should probably be in the Start function.

Reply:
Agree, make more sense to move it into Start function, done.